Add support for TM1638 Led and Key component

skykingjwc opened the pull request:
# What does this implement/fix?

Adds a component for the TM1638 Led and Key PCB


## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**
[https://github.com/esphome/esphome-docs/pull/1991](https://github.com/esphome/esphome-docs/pull/1991)

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml

display:
    platform: tm1638
    id: tm1638_display
    stb_pin: 5
    clk_pin: 18
    dio_pin: 23
    intensity: 5
    update_interval: 5s
    lambda: |-
      it.print("0123");

switch:
   - platform: tm1638
     id: MyLed
     led: 0

#Pressing the left most button will turn on the left most LED.
binary_sensor:
  - platform: tm1638
    name: "My Button"
    id: MyButton
    key: 0
    filters:
      - delayed_on: 10ms
    on_press:
     then:
        - switch.turn_on: MyLed
    on_relase
     then:
        - switch.turn_off: MyLed


```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

mvturnho committed:
setup python codgen for tm1637_binary_sensor

mvturnho committed:
added cpp code for diaplay keys

mvturnho committed:
fix sensor ON Off publish

mvturnho committed:
lint fixes

mvturnho committed:
fixed issue where no binary sensors are added, so no need to compile the TM1637 binary_sensor code

mvturnho committed:
clang

mvturnho committed:
clang again

skykingjwc committed:
Porting from TM1637 PR, any time a method from GPIO is called the chip resets
Example this->clk_pin_->setup();

skykingjwc committed:
Display and time are working

skykingjwc committed:
Leds and buttons working

skykingjwc committed:
Created class for LEDs to operate as switches,  not sure what to do about exposing the SetLed method

skykingjwc committed:
Switche automations working, just need to get the LED to turn on now

skykingjwc committed:
Moved TM1638 files to components folder and reverted unintentional changes to TM1637 code

skykingjwc committed:
Moved TM1638 files to components folder and reverted unintentional changes to TM1637 code

skykingjwc committed:
Commit for ssieb

Commit:
Make it work

skykingjwc committed:
Merge pull request #1 from ssieb/tm1638

make it work

skykingjwc committed:
Removed comments and whitespace

skykingjwc committed:
Remove default intensity value, removed bitdelay, fixed initial intensity

skykingjwc commented:
This isn't for the TM1637.  Not sure why that label was added.

skykingjwc committed:
Removed unused import

ssieb commented on esphome/components/tm1638/tm1638.cpp:
```diff
@@ -0,0 +1,302 @@
+        ESP_LOGI(TAG, "TM1638 String is too long for the display!");
+        break;
+      }
+      this->buffer_[pos] = data;
```
There's a valid compile warning at this line.  `data` won't be initialized if the character is outside of the range in line 175.

skykingjwc replied:
I agree that this needs to be addressed, but I am not seeing a compiler warning when I build using the command
`esphome run ..\tests\TM1638test.yaml  --device COM3`

TM1638test.yaml is the config file I'm using for testing.

How are you getting the warning to display?  I'll want to reproduce it so I can verify that it is fixed.

ssieb replied:
I just did `esphome compile` (or `run`):
```
Compiling .pioenvs/switchtest/src/esphome/components/tm1638/tm1638.cpp.o
src/esphome/components/tm1638/tm1638.cpp: In member function 'uint8_t esphome::tm1638::TM1638Component::print(uint8_t, const char*)':
src/esphome/components/tm1638/tm1638.cpp:198:26: warning: 'data' may be used uninitialized in this function [-Wmaybe-uninitialized]
  198 |       this->buffer_[pos] = data;
      |       ~~~~~~~~~~~~~~~~~~~^~~~~~
```

skykingjwc replied:
Got it.  Fix has been pushed to the repo.
https://github.com/esphome/esphome/pull/3340/commits/cca3408856e91266842173000788c21569400750

skykingjwc replied:
> Hey there @glmnet, mind taking a look at this pull request as it has been labeled with an integration (`tm1637`) you are listed as a [code owner](https://github.com/esphome/esphome/blob/dev/CODEOWNERS#L204) for? Thanks! (message by CodeOwnersMention)

How can we get the TM1637 tag removed?  The TM1637 has nothing to do with this PR.

skykingjwc replied:
I merged from dev, and it seems to have resolved the TM1637 issue

ssieb commented on esphome/components/tm1638/tm1638.cpp:
```diff
@@ -0,0 +1,302 @@
+    }
+
+    if (*str == '.') {
+      if (pos != start_pos)
+        pos--;
+      this->buffer_[pos] |= 0b10000000;  // handles the decimal point or period
```
What happens if there is more than one `.` in a row?  You will collapse them all into a single dot.

skykingjwc replied:
Just pushed a fix.
https://github.com/esphome/esphome/pull/3340/commits/a91bed1c140802ee598e1c44ed7bd5b388405f7d

ssieb reviewed (changes requested):
Why is the `tm1637` directory included in this PR?
I've tested this and it works.

skykingjwc committed:
Bugfix for handling sequential dots in display lambda
Note: This bug also exists in the MAX7219 component which is where a copied the code from

skykingjwc committed:
Fixed typo in comments

skykingjwc committed:
Bugfix for handling unknown characters and null reference for data variable in print function
Removed unecessary comment

skykingjwc committed:
Merge branch 'dev' into TM1638_Component

skykingjwc committed:
Clang

skykingjwc committed:
Black formatter

skykingjwc committed:
Removed unecessary include

skykingjwc committed:
ci-custom

skykingjwc committed:
Added pragma

skykingjwc committed:
Fixed include order

skykingjwc committed:
Clang

skykingjwc committed:
Clang

skykingjwc committed:
CLang

skykingjwc committed:
Clang

skykingjwc committed:
Clang

skykingjwc committed:
Added nolint to buffer creation

skykingjwc committed:
NOLINT

skykingjwc committed:
Clang

skykingjwc committed:
clang

skykingjwc committed:
Bugfix for key listener added test yaml

skykingjwc commented:
Addressed all clang issues, linked documentation PR, added tests for new component in tests folder.

skykingjwc committed:
Merge branch 'esphome:dev' into TM1638_Component

skykingjwc committed:
Delete stackdump file

skykingjwc committed:
Removed unecessary logging

skykingjwc committed:
Added this-> to member and property accessors per sseib

skykingjwc committed:
Missed a this->  added it

skykingjwc committed:
Removed unecessary comments, moved buffer instantiate to class definition

skykingjwc committed:
Removed commented code

skykingjwc committed:
Remove whitespace

skykingjwc committed:
Uncommented last 4 characters in translation table

skykingjwc committed:
Replaced switch with binary output

jesserockz commented on esphome/components/tm1638/output/tm1638_led.cpp:
```diff
@@ -0,0 +1,18 @@
+void TM1638Led::write_state(bool state) { tm1638_->set_led(led_, state); }
+
+void TM1638Led::dump_config() {
+  //LOG_SWITCH("", "TM1638 LED", this);
+  //ASK SSIEB about this
```
```suggestion
  LOG_BINARY_OUTPUT("", "TM1638 LED", this);
```

skykingjwc replied:
Removed comments

jesserockz replied:
You missed the point of my suggestion, it was to change to LOG_BINARY_OUTPUT, not just remove the comments...

skykingjwc replied:
My bad,  must have been getting tired. 
Added LOG_BINARY_OUTPUT
c0f9d8b

skykingjwc committed:
Added switches back to TM1638_Component

skykingjwc committed:
Call publish state after changing a switch led state

skykingjwc committed:
Clang

jesserockz commented on tests/tm1638test.yaml:
```diff
@@ -0,0 +1,165 @@
+esphome:
```
You need to add to an existing test file (1,2,3,4,5), new files are ignored

skykingjwc replied:
Added tests to test5.yaml

jesserockz replied:
Dont forget to delete this yaml file too

skykingjwc replied:
Deleted tm1638test.yaml from repo

ssieb replied:
It's still here.  You need to `git rm` it.

skykingjwc committed:
Added TM1638 Dispalay to test5

skykingjwc committed:
Added empty line to test5

skykingjwc committed:
Added binary output logging

skykingjwc committed:
Added binary output, switches, and binary sensors to test5.yaml

skykingjwc committed:
Remove trailing whitespace

skykingjwc committed:
Clang, LOG_BINARY_OUTPUT too many arguments

jesserockz commented on esphome/components/tm1638/switch/__init__.py:
```diff
@@ -0,0 +1,24 @@
+from esphome.const import CONF_ID, CONF_LED
+from ..display import tm1638_ns, TM1638Component, CONF_TM1638_ID
+
+TM1638SwitchLed = tm1638_ns.class_("TM1638SwitchLed", switch.Switch)
```
```suggestion
TM1638SwitchLed = tm1638_ns.class_("TM1638SwitchLed", switch.Switch, cg.Component)
```

jesserockz commented on esphome/components/tm1638/switch/__init__.py:
```diff
@@ -0,0 +1,24 @@
+        cv.GenerateID(CONF_TM1638_ID): cv.use_id(TM1638Component),
+        cv.Required(CONF_LED): cv.int_range(min=0, max=7),
+    }
+)
```
```suggestion
).extend(cv.COMPONENT_SCHEMA)
```

jesserockz commented on esphome/components/tm1638/switch/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await switch.register_switch(var, config)
```
```suggestion
    var = await switch.new_switch(config)
    await cg.register_component(var, config)
```

jesserockz commented on esphome/components/tm1638/output/__init__.py:
```diff
@@ -0,0 +1,24 @@
+from esphome.const import CONF_ID, CONF_LED
+from ..display import tm1638_ns, TM1638Component, CONF_TM1638_ID
+
+TM1638OutputLed = tm1638_ns.class_("TM1638OutputLed", output.BinaryOutput)
```
```suggestion
TM1638OutputLed = tm1638_ns.class_("TM1638OutputLed", output.BinaryOutput, cg.Component)
```

jesserockz commented on esphome/components/tm1638/output/__init__.py:
```diff
@@ -0,0 +1,24 @@
+        cv.GenerateID(CONF_TM1638_ID): cv.use_id(TM1638Component),
+        cv.Required(CONF_LED): cv.int_range(min=0, max=7),
+    }
+)
```
```suggestion
).extend(cv.COMPONENT_SCHEMA)
```

jesserockz commented on esphome/components/tm1638/output/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await output.register_output(var, config)
```
```suggestion
    await output.register_output(var, config)
    await cg.register_component(var, config)
```

skykingjwc committed:
Update esphome/components/tm1638/switch/__init__.py

Add cg.Component

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

skykingjwc committed:
Added cv.COMPONENT_SCHEMA

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

skykingjwc committed:
added new_switch

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

skykingjwc committed:
added cg,Component

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

skykingjwc committed:
added vc.COMPONENT_SCHEMA

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

skykingjwc committed:
add cg.register_component

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

skykingjwc committed:
Delete tm1638test.yaml

skykingjwc committed:
Removed unecessary import

ssieb commented on esphome/components/tm1638/binary_sensor/__init__.py:
```diff
@@ -0,0 +1,23 @@
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await binary_sensor.register_binary_sensor(var, config)
```
```suggestion
    var = await binary_sensor.new_binary_sensor(config)
```

ssieb commented on esphome/components/tm1638/tm1638.cpp:
```diff
@@ -0,0 +1,288 @@
+  }
+}
+
+void TM1638Component::reset_(bool on_off) {
```
I don't see anywhere that this is called with `true`, so you can remove the parameter.

skykingjwc replied:
Removed unnecessary parameter

skykingjwc committed:
await new_binary_sensor added

Co-authored-by: Samuel Sieb <samuel-github@sieb.net>

skykingjwc committed:
Removed uncessary parameter

skykingjwc committed:
Removed unecessary parameter

skykingjwc committed:
Merge branch 'TM1638_Component' of https://github.com/skykingjwc/esphome into TM1638_Component

skykingjwc committed:
Merge branch 'esphome:dev' into TM1638_Component

skykingjwc committed:
Removed test file

skykingjwc committed:
Merge branch 'TM1638_Component' of https://github.com/skykingjwc/esphome into TM1638_Component

skykingjwc committed:
Updated header file, removed unecessary argument

skykingjwc committed:
Removed unecessary import

ssieb commented:
I will commit after verifying something with the contributor.
